Allow netdata plugins to expose functions for querying more information about specific charts

ktsaou opened the pull request:
Work in progress...

With this PR, plugins are allowed to set custom functions that can be called to provide additional details about anything. For example:

1. `apps.plugin` can provide a real-time `top` like output
2. `ebpf.plugin` can provide a real-time break down of all the traffic coming in and going out of a node
3. `go.d.plugin`, `postgresql` module can provide a real-time updated list of locks as they happen
4. `go.d.plugin`, `mysql` module can provide a real-time updated list of the slow queries of the database server

There are a myriad of possible uses, like getting information about raid arrays status, routing information, firewall information, grepping the last X number of logs that match something, showing application exceptions, etc. The possibilities are endless...

## Goals

- [x] Support for running functions by internal plugins (`C`)
- [x] Support for running functions by external plugins (any language)
- [x] Support for running functions by streaming children (the request to run a function will be automatically routed to the data collecting agent and its plugin, and the reply back to the parent requesting this function execution)
- [x] Support for Web API and ACLK for querying functions

## What is a function?

A function is a `key` - `value` pair. The `key` uniquely identifies the function within a node. The `value` is a function (i.e. code) to be run by a data collector when the function is invoked.

A function has a `format`, depicting the output format that is expected from the function. Possible formats are `csv`, `json`, `text`. The netdata agent does not do anything about the format. It only passes data from the collector to the dashboard. So, formats need to be agreed between the collector and the UI.

Each function is attached to a chart. But many charts can share the same function. When multiple charts share the same function (exactly the same `key`), the dashboard should present it only once.

## Example of a function

`postgresql` module of `go.d.plugin` may define a function called `PostgresqlLocks` which generates a list of the currently held locks.

The function has the ability to find the list of locks at the server level (the whole postgresql instance), or at a given database, or at a given table.

The `postgresql` module of `go.d.plugin` can generate functions like these:

- `PostgresqlLocks job:local` to query all the locks of the `local` data collection job.
- `PostgresqlLocks job:local db:sales` to query the locks of database `sales`.
- `PostgresqlLocks job:local db:sales table:revenue` to query the locks of table `revenue` or database `sales`.

The way this is designed, the plugin can attach the function to multiple places (what is usually called in UI terms: a context sensitive menu). So, it attaches these functions as follows:

- `PostgresqlLocks job:local` is attached to the global locks chart of the server.
- `PostgresqlLocks job:local db:sales` is attached to the locks chart of the database `sales`.
- `PostgresqlLocks job:local db:sales table:revenue` is attached to the locks chart of the table `revenue` of database `sales`.

Now let's assume there are 2 postgresql servers: `db1` and `db2`. At the overview dashboard, when we see the global postgresql locks composite chart we have 2 context sensitive functions we may run:

- `PostgresqlLocks job:local @db1`
- `PostgresqlLocks job:local @db2`

When we go to the locks composite chart of the databases, we may have:

- `PostgresqlLocks job:local db:sales @db1`
- `PostgresqlLocks job:local db:sales @db2`
- `PostgresqlLocks job:local db:customers @db1`
- `PostgresqlLocks job:local db:customers @db2`

When we go the locks composite chart of the tables, we may see:

- `PostgresqlLocks job:local db:sales table:revenue @db1`
- `PostgresqlLocks job:local db:sales table:accounts @db1`
- `PostgresqlLocks job:local db:sales table:contacts @db1`
- `PostgresqlLocks job:local db:sales table:revenue @db2`
- `PostgresqlLocks job:local db:sales table:accounts @db2`
- `PostgresqlLocks job:local db:sales table:contacts @db2`

So, the function is attached to multiple places, with different parameters, so that the user can run them with a single click.

When the user clicks any of them, a new modal opens, which shows:

1. The server the function is running (drop down menu)
2. An edit box for the user to edit the function and its parameters (yes, they are editable, even if something wrong is written there, the function should return with a human readable error explaining what is wrong)
3. A help line describing what other parameters the user can put there (eg. top 10, not all)
4. An auto-refresh frequency (run this every 10 seconds)
5. A timeout drop down in case it times out and user wants to increase it (the default timeout is specified by the plugin itself)
6. The output of the plugin, rendered the way it is supposed according to its format

The output may be interactive, to dynamically add more options to the function itself. For example, `ebpf.plugin` may be used to provide a breakdown of the traffic of the server in a [sankey chart](https://observablehq.com/@d3/sankey) (or any of the hundreds of [available visualizations](https://observablehq.com/@d3/gallery)), to allow interactively drilling down to details.

## Functions in internal plugins

The plugin has to call:

- [x] `rrdset_collector_started()` once, when it starts.
- [x] `rrdset_collector_finished()` once, when it exits - all its functions will automatically become non-queriable after this call. This is automatically called for all threads that are started by netdata agent core.
- [x] `rrdset_collector_add_function()` to attach a function to a chart. Registering a function will fail if `rrdset_collector_started()` has not been called by the same thread that registers the function. This call accepts a callback function that will execute the function in question and generate the output required.

## Functions in external plugins

### Defining functions

The plugin has to output:

```
CHART ...
DIMENSION ...
FUNCTION [GLOBAL]  "{UNIQUE NAME AND PARAMETERS OF FUNCTION}" {TIMEOUT} "{HELP}"
```

- `GLOBAL` is an optional parameter that makes netdata attach this function to the entire host, instead of a chart.
- `UNIQUE NAME AND PARAMETERS OF FUNCTION` uniquely defines the function to be called. If it contains spaces, it must be enclosed in single or double quotes.
- `TIMEOUT` is the expected timeout the collector believes the function should have when executed.
- `HELP` is optional. It is a text describing the command line parameters supported by the function. If it contains spaces, it must be enclosed in single or double quotes.

Examples:

```
FUNCTION text 10 "PostgresqlLocks job:local db:sales" "top:10 deepscan"
```

The above says that there is a function that returns in `text` format, the Locks of PostgreSQL for the database `sales` of the `local` postgresql instance. The function supports the parameters `top:10` (obviously the user can change that to any number) and `deepscan`.

### Executing functions

To support bidirectional communication with the netdata agent, external plugins now have 2 pipes with netdata. The `stdout` like they had before, but also their `stdin`.

When Netdata needs to execute a function from a plugin it sends to its `stdin`:

```
FUNCTION {UNIQUE_ID} {TIMEOUT} {UNIQUE NAME AND PARAMETERS OF FUNCTION} {OPTIONS}
```

- `UNIQUE_ID` is a number generated by netdata core to allow multiplexing / queuing multiple function executions.
- `TIMEOUT` is the timeout in seconds
- `UNIQUE NAME AND PARAMETERS OF FUNCTION` is the function definition as it was given by the plugin when it defined the function.
- `OPTIONS` is similar to command line options of unix command lines. This can be whatever the function supports.

The plugin is expected to to get `UNIQUE NAME AND PARAMETERS OF FUNCTION`, run the relative function and then output to its `stdout` (together with `BEGIN`, `SET`, `END`, but not multiplexed in any particular order):

```
FUNCTION_RESULT_BEGIN {UNIQUE_ID} {STATUS} {CONTENT_TYPE}
# result data
FUNCTION_RESULT_END
```

If the result data is binary, the maximum size is 16k. If the result data is text, each line cannot be more than 16k.

- `UNIQUE_ID` is the unique ID that was sent to the plugin when requesting function execution
- `STATUS` follows the HTTP response codes, so 200 is the OK.
- `CONTENT_TYPE` is `application/json`, `text/plain`, `text/html` or whatever the plugin produces.

Between `FUNCTION_RESULT_BEGIN` and `FUNCTION_RESULT_END` nothing else should be multiplexed. Plugins.d does not parse it. Both of these statements should be alone at their lines (so, a new line is expected before and after them).

## Calling these functions

There are 2 interfaces to call these functions:

- [x] `rrdset_call_function_and_wait()` which executes a function and blocks until the function is executed, or the timeout is expired.
- [x] `rrdset_call_function_async()` which executes a function and returns immediately. Once the call is executed the callback function given will be executed.

## How the UI learns about these functions

- [x] They are returned in the `jsonwrap` of data queries (both context and chart ones).
- [x] New API end point called `/api/v1/function` to execute these functions
- [x] New API end point called `/api/v1/functions` to get a list of all available functions

## How the UI presents these functions to users

There should be a drop down menu on each chart, presenting all the functions that are available to call for this chart. This is tricky, because functions are usually related to something about a specific node (a data collection job, an instance of some data collection entity, etc). So, we need to find an easy way to present them.

---

## Additional improvements

- [x] Streaming protocol was negotiating a version. The minimum version between parent and child was chosen. Now the streaming protocol negotiates capabilities, i.e. individual features that are available on both the parent and the child. This is done in a backwards compatible way, using the same version field, allowing newer children to detect the capabilities of older parents and vice versa.
- [x] Health was previously initialized on each chart and dimension synchronously. Now all health related initialization for charts and dimensions happens by the health thread, allowing busy parents to onboard children in half the time.
- [x]  Plugins.d was using a create/destroy process for updating chart labels. Now the labels are being updated in-place, migrating the live chart labels to the ones required, while adding, updating or removing labels as necessary. This also improves the performance of streaming parents significantly when most charts have labels.
- [x] The election of parents at the child side had a black-hole condition, where the child was repeatedly selecting an unresponsive parent without proceeding to the next ones. Now the parents are always rotated when a socket is established, so all parents are tried in rotation. Tested on a setup of 10 nodes, all in active - active cluster (each one had as destination all the others). They manage to successfully connect to each other no matter what conditions are there.
- [x] This PR includes all the necessary work to have reliable bidirectional streaming communication. So, from now on parents and children can exchange information in realtime.
- [x] This PR fixes unhandled error conditions at the streaming communication, so it should generally be a lot more robust.
- [x] Fixed crashes due to service thread archiving charts while they where collected.
- [x] Streaming sender has been tremendously cleaned up.
- [x] Several optimizations at the cycle BEGIN-SET-END of streaming metrics to increase the performance above 600k metrics / second / streaming connection.

cakrit commented:
- I don't understand the first paragraph of "What is a function". Suggestion: Use the postgresql example to illustrate this point (and others further down).
- From what I see, I don't get if and how a function could accept parameters (e.g. give me the slow queries for the past 20minutes)
- I don't see an addition to the web server, to be able to request execution of these functions. Is this for later?

shyamvalsan commented:
Feedback from a UI perspective.

This is powerful functionality - it shouldn't only be hidden inside a drop down. 

If we support table charts, can we not use this functionality to do, for example, a table of top 10 processes by utilization OR top 10 postgresql tables by bloat %. These tables do not need to be high fidelity (1s) but could be refreshed every 60 seconds or 300 seconds or on page refresh/user click. But this could be a way to get users the info about the top N of a measurement they care about without having to monitor and chart every instance of that measurement type.


For the [problem that this user](https://discord.com/channels/847502280503590932/1022693928874549368) ran into with PostgreSQL collector - where they had 720 tables and 5000+ indexes and Netdata created 50K charts just for Postgres - we could instead only show aggregate charts by default (not per table or per index) but also have tabular charts for:

- top 10 tables by dead row ratio
- top 10 tables by bloat percentage
- top 10 tables by table cache miss ratio
- etc.

And these tables can be refreshed every 5 minutes or even 60 minutes (as long as user can trigger a refresh if they wish to). This would solve this problem (and possibly similar problems for other collectors).

ktsaou commented:
> @ktsaou besides the timeout, is there any max limit on the amount of data that such an interactive query can return? is there such a thing as too much data to try and fetch interactively?

If the response is text based, 16k per line is the only requirement for the compression protocol to work between parents and children. I don't see any other hard limitation.

ktsaou commented:
> * I don't understand the first paragraph of "What is a function". Suggestion: Use the postgresql example to illustrate this point (and others further down).
> * From what I see, I don't get if and how a function could accept parameters (e.g. give me the slow queries for the past 20minutes)
> * I don't see an addition to the web server, to be able to request execution of these functions. Is this for later?

Added all of them @cakrit

ktsaou commented:
> Feedback from a UI perspective.
> 
> This is powerful functionality - it shouldn't only be hidden inside a drop down.
> 
> If we support table charts, can we not use this functionality to do, for example, a table of top 10 processes by utilization OR top 10 postgresql tables by bloat %. These tables do not need to be high fidelity (1s) but could be refreshed every 60 seconds or 300 seconds or on page refresh/user click. But this could be a way to get users the info about the top N of a measurement they care about without having to monitor and chart every instance of that measurement type.
> 
> For the [problem that this user](https://discord.com/channels/847502280503590932/1022693928874549368) ran into with PostgreSQL collector - where they had 720 tables and 5000+ indexes and Netdata created 50K charts just for Postgres - we could instead only show aggregate charts by default (not per table or per index) but also have tabular charts for:
> 
> * top 10 tables by dead row ratio
> * top 10 tables by bloat percentage
> * top 10 tables by table cache miss ratio
> * etc.
> 
> And these tables can be refreshed every 5 minutes or even 60 minutes (as long as user can trigger a refresh if they wish to). This would solve this problem (and possibly similar problems for other collectors).

@shyamvalsan I updated the description at the top with more info. My goal is not to restrict it, but to allow users use hundreds or even thousands of such ready to use functions with a single click. To support something like "context sensitive menu", so to have the right functions at the right place.

The user should always have the flexibility to customize each function the way he/she sees fit.

sashwathn commented:
@cakrit : Now that we allow custom labels at the job level, I want to understand the actual use of this.
I believe what we are allowing users to set is some custom labels for their jobs (to differentiate one from another). But we already have a default label _collect_job which we are able to "Group by" and "Filter" from and these custom labels will now be an additional label over and above these and I am not sure if these are good from a UX perspective.

@shyamvalsan and @ilyam8 for more comments.

hugovalente-pm commented:
> This is powerful functionality - it shouldn't only be hidden inside a drop down.

I agree with @shyamvalsan here, this is really a very powerful and helpful feature to help troubleshoot. Making sure users know this is available is key

> If we support table charts, can we not use this functionality to do, for example, a table of top 10 processes by utilization OR top 10 postgresql tables by bloat %. These tables do not need to be high fidelity (1s) but could be refreshed every 60 seconds or 300 seconds or on page refresh/user click. But this could be a way to get users the info about the top N of a measurement they care about without having to monitor and chart every instance of that measurement type.

@shyamvalsan @ktsaou the question here is how costly will it be to execute these functions with a given frequency ?
I believe initial idea was to have them triggered per user request on a chart (or other place were he could fire them), if we put them and allow users to configure table charts with them, where they can configure whoever they want we can have some heavy scenarios no? 
to try to give an example, top 50 indexes by bloat percentage across all the dbs being monitored on a node updated every 15 sec. (or something heavier)

hugovalente-pm commented:
> From what I see, I don't get if and how a function could accept parameters (e.g. give me the slow queries for the past 20minutes)
@cakrit just to check and have a common understanding on the timeframe of requests, afaik these functions will be executed on run-time you can not specify a time interval to get the results of X function on the past Y minutes/hours

is this right @ktsaou ?

ktsaou commented:
Guys, these functions are like unix command lines.

Let's assume we have a command / function named `PostgresqlSlowQuery`. When you run it, it returns the top 10 of currently running slow queries of a postgresql instance.

But then you can say `PostgresqlSlowQuery top:20` to get the top 20 of them.

If the algorithm supports showing the slow queries of the last 20 minutes, we could state it like `PostgresqlSlowQuery mins:20`.

The idea is that the command / function can support an unlimited number of parameters. Anything the developer of the function wants or believes it is going to be useful. He also supplies a help string to allow users remember them.

Then this function, with specific parameters, can be attached to any chart, as quick access for users use.

But the users can run it any way they like it. The function behind the scenes is attached to the node and the specific collector that executes it. So, we can give to users all the flexibility they need at the UI.

hugovalente-pm commented:
Ok, thanks for confirming that @ktsaou it will really depend on the function.

What about the performance question raised about having the functions executing on a specific internal to fill a table? (Comment link below)

https://github.com/netdata/netdata/pull/13720#issuecomment-1259352887

You're saying users should be aware and careful of the configurations they do?

vkalintiris commented on database/rrdfunctions.h:
```diff
@@ -0,0 +1,19 @@
+extern void rrd_collector_started(void);
+extern void rrd_collector_finished(void);
+
+extern void rrd_collector_add_function(RRDSET *st, const char *name, const char *help, const char *format, int timeout, bool sync, int (*function)(BUFFER *wb, int timeout, const char *function, void *collector_data, void (*callback)(BUFFER *wb, int code, void *callback_data), void *callback_data), void *collector_data);
```
@ktsaou it's worthwhile `typedef`-ing callbacks to make the code more concise and easier to skim through.

thiagoftsm commented:
Hello @ktsaou,

I am starting review your PR and I will address your suggestion for ebpf in another PR.

Best regards

thiagoftsm commented on libnetdata/popen/popen.c:
```diff
@@ -89,118 +137,175 @@ static inline void convert_argv_to_string(char *dst, size_t size, const char *sp
         // posix_spawnattr_init() succeeded
         if (posix_spawnattr_destroy(&attr))
-            error("posix_spawnattr_destroy");
+            error("POPEN: posix_spawnattr_destroy() failed");
```
I ran coverity on this PR and it is reporting:

```sh
*** CID 357335:  Resource leaks  (RESOURCE_LEAK)
/libnetdata/popen/popen.c: 281 in popene_internal()
275                 error("POPEN: posix_spawnattr_destroy() failed");
276         }
277     
278         if (posix_spawn_file_actions_destroy(&fa))
279             error("POPEN: posix_spawn_file_actions_destroy() failed");
280     
>>>     CID 357335:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "fp_child_stdout" going out of scope leaks the storage it points to.
281         return ret;
282     
283     error_after_posix_spawn_file_actions_init:
284         if (posix_spawn_file_actions_destroy(&fa))
285             error("POPEN: posix_spawn_file_actions_destroy() failed");
286     
```

thiagoftsm replied:
We have another one:

```sh
*** CID 357337:  Resource leaks  (RESOURCE_LEAK)
/libnetdata/popen/popen.c: 281 in popene_internal()
275                 error("POPEN: posix_spawnattr_destroy() failed");
276         }
277     
278         if (posix_spawn_file_actions_destroy(&fa))
279             error("POPEN: posix_spawn_file_actions_destroy() failed");
280     
>>>     CID 357337:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "fp_child_stdin" going out of scope leaks the storage it points to.
281         return ret;
282     
283     error_after_posix_spawn_file_actions_init:
284         if (posix_spawn_file_actions_destroy(&fa))
285             error("POPEN: posix_spawn_file_actions_destroy() failed");
286     
```

ktsaou replied:
I think this is by design. We need it to leak out of the function it is defined.
Do you see any problem with it?

thiagoftsm replied:
I agree with you, we can mark this as a false flag. 
I copied it here for we do no forget to mark it.

ktsaou replied:
If found it. It leaks both fp and the fpp params are NULL.

ktsaou replied:
I think I fixed it @thiagoftsm - can you pass it again over coverity?

thiagoftsm commented on streaming/receiver.c:
```diff
@@ -762,7 +797,8 @@ static int rrdpush_receive(struct receiver_state *rpt)
 
     // cleanup
-    fclose(fp);
+    fclose(fp_in);
```
For these lines:

```sh
*** CID 357334:  Null pointer dereferences  (NULL_RETURNS)
/streaming/receiver.c: 801 in rrdpush_receive()
795             netdata_mutex_unlock(&rpt->host->receiver_lock);
796             rrd_unlock();
797         }
798     
799         // cleanup
800         fclose(fp_in);
>>>     CID 357334:  Null pointer dereferences  (NULL_RETURNS)
>>>     Dereferencing a pointer that might be "NULL" "fp_out" when calling "fclose".
801         fclose(fp_out);
802         return (int)count;
803     }
804     
805     void *rrdpush_receiver_thread(void *ptr) {
806         netdata_thread_cleanup_push(rrdpush_receiver_thread_cleanup, ptr);

** CID 357333:  Insecure data handling  (TAINTED_SCALAR)
```

ktsaou replied:
fixed it - will merge in a while

ktsaou replied:
It is merged.

thiagoftsm commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -4134,6 +4137,68 @@ static int check_capabilities() {
+                      timeout_s?timeout_s:"(unset)",
+                      function?function:"(unset)");
+            }
+            else {
```
Another coverity for these lines:

```sh
*** CID 357333:  Insecure data handling  (TAINTED_SCALAR)
/collectors/apps.plugin/apps_plugin.c: 4163 in reader_main()
4157                           words[0],
4158                           transaction?transaction:"(unset)",
4159                           timeout_s?timeout_s:"(unset)",
4160                           function?function:"(unset)");
4161                 }
4162                 else {
>>>     CID 357333:  Insecure data handling  (TAINTED_SCALAR)
>>>     Passing tainted expression "*timeout_s" to "str2i", which uses it as a loop boundary.
4163                     int timeout = str2i(timeout_s);
4164                     if(timeout <= 0) timeout = PLUGINS_FUNCTIONS_TIMEOUT_DEFAULT;
4165     
4166                     internal_error(true, "FUNCTION '%s' with transaction id '%s' received", function, transaction);
4167     
4168                     netdata_mutex_lock(&mutex);

** CID 357339:  Concurrent data access violations  (MISSING_LOCK)
/libnetdata/popen/popen.c: 82 in netdata_popen_tracking_init()
```

ktsaou replied:
I made all str2x functions have a maximum string they parse.

thiagoftsm commented on libnetdata/popen/popen.c:
```diff
@@ -2,81 +2,129 @@
+//
+// Callers should call myp_free() to clean up resources.
+void netdata_popen_tracking_init(void) {
+    info("process tracking enabled.");
```
For these lines we have:

```sh
*** CID 357339:  Concurrent data access violations  (MISSING_LOCK)
/libnetdata/popen/popen.c: 82 in netdata_popen_tracking_init()
76     // which will be handled internally, by calling myp_reap, to
77     // avoid issues with already reaped processes during wait calls.
78     //
79     // Callers should call myp_free() to clean up resources.
80     void netdata_popen_tracking_init(void) {
81         info("process tracking enabled.");
>>>     CID 357339:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "netdata_popen_tracking_enabled" without holding lock "netdata_popen_tracking_mutex". Elsewhere, "netdata_popen_tracking_enabled" is accessed with "netdata_popen_tracking_mutex" held 1 out of 2 times (1 of these accesses strongly imply that it is necessary).
82         netdata_popen_tracking_enabled = true;
83     
84         if (netdata_mutex_init(&netdata_popen_tracking_mutex) != 0)
85             fatal("netdata_popen_tracking_init() mutex init failed.");
86     }
87     

** CID 357338:  Resource leaks  (RESOURCE_LEAK)
/streaming/receiver.c: 710 in rrdpush_receive()
```

ktsaou replied:
This seems a false positive - it is the init function - the mutex is not initialized yet.

thiagoftsm commented on streaming/receiver.c:
```diff
@@ -667,8 +701,9 @@ static int rrdpush_receive(struct receiver_state *rpt)
+    FILE *fp_in = fdopen(rpt->fd, "r");
+    if(!fp_in) {
         log_stream_connection(rpt->client_ip, rpt->client_port, rpt->key, rpt->host->machine_guid, rrdhost_hostname(rpt->host), "FAILED - SOCKET ERROR");
         error("STREAM %s [receive from [%s]:%s]: failed to get a FILE for FD %d.", rrdhost_hostname(rpt->host), rpt->client_ip, rpt->client_port, rpt->fd);
```
```sh
*** CID 357338:  Resource leaks  (RESOURCE_LEAK)
/streaming/receiver.c: 710 in rrdpush_receive()
704         FILE *fp_out = fdopen(rpt->fd, "w");
705         FILE *fp_in = fdopen(rpt->fd, "r");
706         if(!fp_in) {
707             log_stream_connection(rpt->client_ip, rpt->client_port, rpt->key, rpt->host->machine_guid, rrdhost_hostname(rpt->host), "FAILED - SOCKET ERROR");
708             error("STREAM %s [receive from [%s]:%s]: failed to get a FILE for FD %d.", rrdhost_hostname(rpt->host), rpt->client_ip, rpt->client_port, rpt->fd);
709             close(rpt->fd);
>>>     CID 357338:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "fp_out" going out of scope leaks the storage it points to.
710             return 0;
711         }
712     
713         rrdhost_wrlock(rpt->host);
714     /* if(rpt->host->connected_senders > 0) {
715             rrdhost_unlock(rpt->host);

** CID 357337:  Resource leaks  (RESOURCE_LEAK)
/libnetdata/popen/popen.c: 281 in popene_internal()
```

ktsaou replied:
I think this is also by design. We need it to go out of the function.

thiagoftsm replied:
I ran coverity again, and it did not find issues. I am going to write a report with tests I ran and approve the PR.

thiagoftsm commented:
I will rebase here locally and proceed with tests. :handshake:

thiagoftsm reviewed (dismissed):
After all coverity report to be fixed, I tested this PR on the following environment:

| Host | is parent? | Is this PR? |
|--------|---------------|------------------|
| Slackware current X86_64 | yes | yes |
|FreeBSD 13.1-RELEASE-p2 Libvirt VM| no | yes |
| Ubuntu 18.04 Libvirt VM| no | no| 
| Raspberry PI Armvl7 | no | no|

The parent was compiled with `libasan` and I did not have error reports.
Children were compiled direct and I did not coredumps.
Cloud worked as expected for all nodes.

LGTM!

thiagoftsm commented:
@ktsaou after review last commits, while I was testing I observed that on parent:

```sh
bash-5.2# ps aux| grep netdata
netdata  16784  0.0  0.0   7140  3968 ?        SN   22:54   0:01 bash /usr/libexec/netdata/plugins.d/tc-qos-helper.sh 1
root     20823  0.0  0.0   6840  2836 pts/4    S+   23:26   0:00 grep netdata
netdata  27203  4.6  4.9 21475636340 799976 ?  SNl  21:54   4:20 netdata
netdata  27207  0.0  0.1 21474970864 25028 ?   SNl  21:54   0:00 /usr/sbin/netdata --special-spawn-server
netdata  27453  0.1  0.2 770444 46120 ?        SNl  21:54   0:08 /usr/libexec/netdata/plugins.d/go.d.plugin 1
netdata  27473  0.2  0.3 262492 60800 ?        SNl  21:54   0:12 /usr/bin/python3 /usr/libexec/netdata/plugins.d/python.d.plugin 1
root     27487  0.1  1.8 21475229768 299672 ?  SNl  21:54   0:07 /usr/libexec/netdata/plugins.d/ebpf.plugin 1
netdata  27490  0.0  0.0 21474877052 8448 ?    SN   21:54   0:00 /usr/libexec/netdata/plugins.d/perf.plugin 1 cycles instructions emulation L1I
netdata  27502  2.1  0.9 21474888036 150260 ?  SNl  21:54   2:01 /usr/libexec/netdata/plugins.d/apps.plugin 1
root     27504  0.0  0.0 21474877112 8648 ?    SN   21:54   0:00 /usr/libexec/netdata/plugins.d/nfacct.plugin 1
netdata  27821  0.0  0.0      0     0 ?        ZN   21:54   0:00 [bash] <defunct>
netdata  30269  0.0  0.0      0     0 ?        ZN   22:06   0:00 [bash] <defunct>
netdata  32397  0.1  0.3 262468 60816 ?        SNl  19:10   0:17 /usr/bin/python3 /usr/libexec/netdata/plugins.d/python.d
``` 

and on child 

```sh
[root@archlinux netdata]# ps aux| grep netdata
netdata    32291  0.8  9.3 520436 92836 ?        SNl  21:54   0:46 netdata
netdata    32293  0.0  1.2  85692 12908 ?        SNl  21:54   0:00 /usr/bin/netdata --special-spawn-server
netdata    32435  0.1  6.0 770440 59848 ?        SNl  21:54   0:07 /usr/libexec/netdata/plugins.d/go.d.plugin 1
root       32454  0.0  6.2 1272684 62128 ?       SNl  21:54   0:03 /usr/libexec/netdata/plugins.d/ebpf.plugin 1
netdata    32455  0.6  1.0 133364 10168 ?        SNl  21:54   0:37 /usr/libexec/netdata/plugins.d/apps.plugin 1
netdata    32677  0.0  0.0      0     0 ?        ZN   21:55   0:00 [bash] <defunct>
netdata    32714  0.0  0.0      0     0 ?        ZN   22:06   0:00 [bash] <defunct>
netdata    32854  0.0  0.3   4544  3316 ?        RN   22:54   0:00 bash /usr/libexec/netdata/plugins.d/tc-qos-helper.sh 1
root       32964  0.0  0.2   3976  2268 pts/0    S+   23:29   0:00 grep netdata

```

Netdata is keeping some `<defunct>` processes. 

Parent was `Slackware current` and child was `Arch Linux`.

ktsaou commented:
Nice catch @thiagoftsm !
Fixed it.

thiagoftsm commented:
@ktsaou I am testing last commits, and I observed these warning while I was compiling:

```sh
[/home/thiago/Netdata/tests_netdata]# make -j4 
collectors/apps.plugin/apps_plugin.c: In function 'apps_plugin_function_process_tree':
collectors/apps.plugin/apps_plugin.c:4283:17: warning: format '%d' expects argument of type 'int', but argument 9 has type 'uid_t' {aka 'unsigned int'} [-Wformat=]
 4283 |                 "<tr>"
      |                 ^~~~~~
......
 4300 |                 , p->user_target?p->user_target->name:"-", p->uid
      |                                                            ~~~~~~
      |                                                             |
      |                                                             uid_t {aka unsigned int}
collectors/apps.plugin/apps_plugin.c:4289:27: note: format string is defined here
 4289 |                 "<td>%s (%d)</td>"  // user
      |                          ~^
      |                           |
      |                           int
      |                          %d
collectors/apps.plugin/apps_plugin.c:4283:17: warning: format '%d' expects argument of type 'int', but argument 11 has type 'gid_t' {aka 'unsigned int'} [-Wformat=]
 4283 |                 "<tr>"
      |                 ^~~~~~
......
 4301 |                 , p->group_target?p->group_target->name:"-", p->gid
      |                                                              ~~~~~~
      |                                                               |
      |                                                               gid_t {aka unsigned int}
collectors/apps.plugin/apps_plugin.c:4290:27: note: format string is defined here
 4290 |                 "<td>%s (%d)</td>"  // group
      |                          ~^
      |                           |
      |                           int
      |                          %d
 OK  ''

```

I used the following CFLAGS:

```sh
CFLAGS="-O1 -ggdb -Wall -Wextra -fsanitize=address -static-libasan -fno-omit-frame-pointer -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1 -Wformat-truncation=2 -Wunused-result"
```

ktsaou commented:
@thiagoftsm check again. That code has been replaced.

Guys, I updated swagger, documentation and fixed all the bugs reported.
I don't have any pending items about this PR.

We target to merge on Monday.

ktsaou commented:
And hey, you can check the 1st implemented function:

```
curl 'http://localhost:19999/api/v1/function?function=processes%20help'
```

`%20` is the space. You need to put them like that in the URL.

To see the output:

```
curl 'http://localhost:19999/api/v1/function?function=processes' | jq .
```

It will give you a lot.

To see the output for process group `netdata` run:

```
curl 'http://localhost:19999/api/v1/function?function=processes%20category:netdata' | jq .
```

To see the output for process named `netdata` run:

```
curl 'http://localhost:19999/api/v1/function?function=processes%20process:netdata' | jq .
```

Yes, `apps.plugin` responds...

ktsaou commented:
Today in master there is this code:

https://github.com/netdata/netdata/blob/7e29d5a90896c777d823c46a232fb0edf4f4525e/libnetdata/popen/popen.c#L336-L344

I added this code back in 2016, due to problems with musl libc.

However the workaround I provided back then is problematic. The reason is that file descriptors are allocated from the minimum number upwards, so this double close of the file descriptor sometimes leads to closing another file descriptor that netdata happened to open while this was running.

I was misled by this and I did it a few more times in the code. And now that netdata manages really a lot of sockets, this little thing became a big problem, suddenly closing file descriptors of other streaming connections or plugins that happen to be initializing between these lines.

I removed it, of course.

Still the question remains: does now netdata work as expected under musl libc? Are plugins exiting properly under musl libc?

thiagoftsm commented:
All right @ktsaou ,

Tomorrow I will dedicate my whole day to test your PR in different scenarios, I am compiling it right now, and I will run coverity on it after this. :handshake: 

Best regards!

thiagoftsm commented on streaming/sender.c:
```diff
@@ -507,14 +489,13 @@ if(!s->rrdpush_compression)
     http[received] = '\0';
     debug(D_STREAM, "Response to sender from far end: %s", http);
-    int32_t version = (int32_t)parse_stream_version(host, http);
+    int32_t version = parse_stream_version(host, http);
```
@ktsaou the unique report from Coverity is this:

```sh
*** CID 357382:  Insecure data handling  (TAINTED_SCALAR)
/streaming/sender.c: 492 in rrdpush_sender_thread_connect_to_parent()
486             rrdpush_sender_thread_close_socket(host);
487             return 0;
488         }
489     
490         http[received] = '\0';
491         debug(D_STREAM, "Response to sender from far end: %s", http);
>>>     CID 357382:  Insecure data handling  (TAINTED_SCALAR)
>>>     Passing tainted expression "http" to "parse_stream_version", which uses it as a loop boundary.
492         int32_t version = parse_stream_version(host, http);
493         if(version == -1) {
494             worker_is_busy(WORKER_SENDER_JOB_DISCONNECT_BAD_HANDSHAKE);
495             error("STREAM %s [send to %s]: server is not replying properly (is it a netdata?).", rrdhost_hostname(host), s->connected_to);
496             rrdpush_sender_thread_close_socket(host);
497             //catch other reject reasons and force to check other destinations
```

, but we are setting the last character as `NULL` few lines before we call the function, and inside function we do not modify it.

thiagoftsm commented:
A quick update, I ran this PR on three different hosts during few hours:

| Host | Distribution | Is Parent |
|--------|-----------------|-------------|
|X86_64 | Slackware Current | Yes |
|X86_64 libvirt VM | Arch Linux | No |
| Armvl7 | Raspberry PI |  No | 

I did not have core dump and PR worked as expected. One of hosts also raised an alert.
In few hours I will run a deeper test for on all areas.

Best regards!

ktsaou commented:
@thiagoftsm I think I fixed the issue with tainted data. Please try again coverity.

thiagoftsm commented:
Hello @ktsaou ,

After to run the [script](https://github.com/netdata/netdata/tree/master/tests/urls) to access our API and verify that everything was ok, I query `https://localhost:19999/api/v1/function?function=processes%20help` to verify what could happen, and my host crashed with the following error:

```sh
2022-10-02 14:13:53: netdata INFO  : PLUGINSD[python.d] : (0919@database/rrdset.c   :rrdset_next_use): MINIMUM MICROSECONDS DELTA of thread 28318 increased from 223 to 416 (+193)
2022-10-02 14:23:53: netdata INFO  : PLUGINSD[python.d] : (0919@database/rrdset.c   :rrdset_next_use): MINIMUM MICROSECONDS DELTA of thread 28318 increased from 416 to 479 (+63)
2022-10-02 14:29:53: netdata INFO  : PLUGINSD[python.d] : (0919@database/rrdset.c   :rrdset_next_use): MINIMUM MICROSECONDS DELTA of thread 28318 increased from 479 to 704 (+225)
2022-10-02 14:30:53: netdata INFO  : PLUGINSD[python.d] : (0919@database/rrdset.c   :rrdset_next_use): MINIMUM MICROSECONDS DELTA of thread 28318 increased from 704 to 749 (+45)
2022-10-02 14:32:53: netdata INFO  : PLUGINSD[python.d] : (0919@database/rrdset.c   :rrdset_next_use): MINIMUM MICROSECONDS DELTA of thread 28318 increased from 749 to 837 (+88)
2022-10-02 14:35:53: netdata INFO  : PLUGINSD[python.d] : (0919@database/rrdset.c   :rrdset_next_use): MINIMUM MICROSECONDS DELTA of thread 28318 increased from 837 to 854 (+17)
2022-10-02 14:36:27: netdata ERROR : PLUGINSD[apps] : (0650@collectors/plugins.d:pluginsd_functi): got a FUNCTION_RESULT_BEGIN without providing the required data (key = '99135dce-425f-11ed-bb43-f82819901f7f', status = '200', format = 'text/plain', expires = '(unset)').
2022-10-02 14:36:27: apps.plugin FATAL : APPS_READER : (4664@collectors/apps.plug:reader_main    ): Received error on read pipe. # : Success

EOF found in spawn pipe.
Shutting down spawn server event loop.
Shutting down spawn server loop complete.
/usr/libexec/netdata/plugins.d/tc-qos-helper.sh: line 233: echo: write error: Broken pipe
```

I ran this test using `[db].mode = ram`

thiagoftsm commented:
I finish the [cited](https://github.com/netdata/netdata/pull/13720#issuecomment-1264431950) tests, and I can confirm that with exception of [this](https://github.com/netdata/netdata/pull/13720#issuecomment-1264659629) reported error, other links worked as expected.

thiagoftsm commented:
> Still the question remains: does now netdata work as expected under musl libc? Are plugins exiting properly under musl libc?

I also faced limitations with `musl` in recent years, but all static binaries are build using it (ping @Ferroin). In theory we are able to compile using `musl`, but we still have some limits imposed by library, for example, we cannot use `dlopen`.

ktsaou commented:
@thiagoftsm by bad... I did some of the changes required...

I fixed both of them (apps.plugin not responding properly on help, netdata crashing on invalid response).

thiagoftsm commented:
> @thiagoftsm by bad... I did some of the changes required...
> 
> I fixed both of them (apps.plugin not responding properly on help, netdata crashing on invalid response).

Thanks, I am stopping both hosts that I was testing, and I will retest with last commits. :handshake:

thiagoftsm commented:
Hello @ktsaou,

I recompile last three commits and I can confirm `netdata` is not crashing anymore. I will proceed with tests.

thiagoftsm commented on collectors/apps.plugin/apps_plugin.c:
```diff
@@ -4134,6 +4174,511 @@ static int check_capabilities() {
+
+    fwrite(buffer_tostring(func_processes_fields), buffer_strlen(func_processes_fields), 1, stdout);
+
+    NETDATA_DOUBLE cpu_divisor = (NETDATA_DOUBLE)(time_factor * RATES_DETAIL / 100);
```
Hello @ktsaou ,

A small report from coverity:

```sh
>>>     CID 358247:  Incorrect expression  (UNINTENDED_INTEGER_DIVISION)
>>>     Dividing integer expressions "time_factor * 10000ULL" and "100ULL", and then converting the integer quotient to type "double". Any remainder, or fractional part of the quotient, is ignored.
4493         NETDATA_DOUBLE cpu_divisor = (NETDATA_DOUBLE)(time_factor * RATES_DETAIL / 100);
```

ktsaou replied:
I solved it.

ktsaou commented:
Working on a UI:
![processes](https://user-images.githubusercontent.com/2662304/193592693-2b5f940d-2e41-4ff9-a488-0011a2537471.gif)

andrewm4894 commented:
instead of each pie chart i wonder if a [horizontal bar chart](https://www.google.com/search?q=horizontal+bar+chart) might be clearer (and could also be "scrollable" if needed to have 100 of entries for example)

andrewm4894 commented:
could also consider adding shading or some sort of glyph to each col of interest in the table - sort of similar to this - so i can see that chrome although is not using much cpu is using decent amount of mem:

tricky to do this and still keep it all clean looking though but could be useful maybe in the table view.

 
![image](https://user-images.githubusercontent.com/2178292/193596485-a996d03a-54c5-4f64-ad77-62afddec1433.png)

thiagoftsm commented:
Hello @ktsaou ,

I decided to test your PR on distributions using old compilers, and on `CentOS 7.9` I am getting these warnings:

```sh
[/home/thiago/Netdata/netdata]# make -j1 
streaming/sender.c:239:9: warning: initializer element is not a constant expression [enabled by default]
         .length = strlen(START_STREAMING_PROMPT_VN),
         ^
streaming/sender.c:239:9: warning: (near initialization for 'stream_responses[0].length') [enabled by default]
streaming/sender.c:248:9: warning: initializer element is not a constant expression [enabled by default]
         .length = strlen(START_STREAMING_PROMPT_V2),
         ^
streaming/sender.c:248:9: warning: (near initialization for 'stream_responses[1].length') [enabled by default]
streaming/sender.c:257:9: warning: initializer element is not a constant expression [enabled by default]
         .length = strlen(START_STREAMING_PROMPT_V1),
         ^
streaming/sender.c:257:9: warning: (near initialization for 'stream_responses[2].length') [enabled by default]
streaming/sender.c:266:9: warning: initializer element is not a constant expression [enabled by default]
         .length = strlen(START_STREAMING_ERROR_SAME_LOCALHOST),
         ^
streaming/sender.c:266:9: warning: (near initialization for 'stream_responses[3].length') [enabled by default]
streaming/sender.c:275:9: warning: initializer element is not a constant expression [enabled by default]
         .length = strlen(START_STREAMING_ERROR_ALREADY_STREAMING),
         ^
streaming/sender.c:275:9: warning: (near initialization for 'stream_responses[4].length') [enabled by default]
streaming/sender.c:284:9: warning: initializer element is not a constant expression [enabled by default]
         .length = strlen(START_STREAMING_ERROR_NOT_PERMITTED),
         ^
streaming/sender.c:284:9: warning: (near initialization for 'stream_responses[5].length') [enabled by default]
 OK  ''
```

when  compile with `gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)`.

The warnings did not affect netdata behavior, I am having alerts, plugins and streaming working as expected on `CentOS`.

thiagoftsm commented:
I finished a round of tests with alerts, and everything is working as expected.

thiagoftsm reviewed (changes requested):
Hello @ktsaou ,

This time I found a different issue, and I decided to request a change instead only comment. When I was compiling on `FreeBSD 13.1-RELEASE-p2` I got this error:

```sh
1 warning generated.
streaming/sender.c:1072:75: error: use of undeclared identifier 'F_GETPIPE_SZ'
    int pipe_buffer_size = fcntl(s->host->rrdpush_sender_pipe[PIPE_READ], F_GETPIPE_SZ);
                                                                          ^
1 error generated.
gmake[2]: *** [Makefile:6156: streaming/sender.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [Makefile:7283: all-recursive] Error 1
gmake: *** [Makefile:3738: all] Error 2
 FAILED  ''
```

Neither the manual installed with FreeBSD nor [online](https://www.freebsd.org/cgi/man.cgi?query=fcntl&sektion=2)  help has `F_GETPIPE_SZ`, probably this is also affecting Apple OS, please, change `flag` for these OS.

ktsaou commented:
I fixed them all @thiagoftsm 

For the `strlen()` in string literals, newer compilers treat it as `sizeof() - 1`. I did the same.
For the `F_GETPIPE_SZ` the code fall back to the old behavior (10k buffer for the pipe).

thiagoftsm commented:
> I fixed them all @thiagoftsm
> 
> For the `strlen()` in string literals, newer compilers treat it as `sizeof() - 1`. I did the same. For the `F_GETPIPE_SZ` the code falls back to the old behavior (10k buffer for the pipe).

I am going to another round of tests on `CentOS` and `FreeBSD`, everything going fine I will approve the PR. :handshake:

thiagoftsm commented:
@stelfrag and @ktsaou ,

It is not related to this PR, but while I was testing it the issue appear again. I cleanup everything inside `/var/cache/netdata` and I started tests, after one hour the tests were stopped with this:

```sh
=================================================================
==9907==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7fa8ecf22900 at pc 0x00000046f56b bp 0x7fa8ecf22010 sp 0x7fa8ecf217d0
WRITE of size 56 at 0x7fa8ecf22900 thread T46
    #0 0x46f56a in unpoison_tm(void*, __sanitizer::__sanitizer_tm*) (/usr/sbin/netdata+0x46f56a)
    #1 0x478a40 in __interceptor_localtime_r (/usr/sbin/netdata+0x478a40)
    #2 0x567939 in log_date libnetdata/log/log.c:464
    #3 0x567939 in info_int libnetdata/log/log.c:769
    #4 0x9d2aa5 in release_statement database/sqlite/sqlite_functions.c:138
    #5 0x7fa90c38ae20 in __GI___nptl_deallocate_tsd (/lib64/libc.so.6+0x88e20)
    #6 0x7fa90c38de44 in start_thread (/lib64/libc.so.6+0x8be44)
    #7 0x7fa90c4189eb in clone3 (/lib64/libc.so.6+0x1169eb)

Address 0x7fa8ecf22900 is located in stack of thread T46 at offset 64 in frame
    #0 0x5677d0 in info_int libnetdata/log/log.c:751

  This frame has 4 object(s):
    [32, 40) 't' (line 460)
    [64, 120) 'tmbuf' (line 461) <== Memory access at offset 64 is inside this variable
    [160, 184) 'args' (line 752)
    [224, 250) 'date' (line 768)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
Thread T46 created by T44 here:
    #0 0x43f5e6 in __interceptor_pthread_create (/usr/sbin/netdata+0x43f5e6)
    #1 0x7fa90c6dd618 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/usr/lib64/libstdc++.so.6+0xdd618)
    #2 0x91b504 in ml_new_host ml/ml.cc:55
    #3 0x9b3372 in rrdhost_create database/rrdhost.c:542
    #4 0x9b4781 in rrdhost_find_or_create database/rrdhost.c:685
    #5 0xbec9d2 in rrdpush_receive streaming/receiver.c:548
    #6 0xbef1a2 in rrdpush_receiver_thread streaming/receiver.c:834
    #7 0x586a8b in thread_start libnetdata/threads/threads.c:199
    #8 0x7fa90c38dfe9 in start_thread (/lib64/libc.so.6+0x8bfe9)

Thread T44 created by T16 here:
    #0 0x43f5e6 in __interceptor_pthread_create (/usr/sbin/netdata+0x43f5e6)
    #1 0x5870eb in netdata_thread_create libnetdata/threads/threads.c:213
    #2 0xbdce19 in rrdpush_receiver_thread_spawn streaming/rrdpush.c:902
    #3 0xc08d2a in web_client_process_request web/server/web_client.c:1469
    #4 0xc0fe18 in web_server_rcv_callback web/server/static/static-threaded.c:304
    #5 0x57ec15 in poll_process_tcp_read libnetdata/socket/socket.c:1493
    #6 0x57ec15 in poll_events libnetdata/socket/socket.c:1812
    #7 0xc0f781 in socket_listen_main_static_threaded_worker web/server/static/static-threaded.c:427
    #8 0xc1262f in socket_listen_main_static_threaded web/server/static/static-threaded.c:546
    #9 0x586a8b in thread_start libnetdata/threads/threads.c:199
    #10 0x7fa90c38dfe9 in start_thread (/lib64/libc.so.6+0x8bfe9)
```

I compiled with the following `CFLAGS`:

```sh
CFLAGS="-O1 -ggdb -Wall -Wextra -fsanitize=address -static-libasan -fno-omit-frame-pointer -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1 -Wformat-truncation=2 -Wunused-result"
```

The test was to compile `netdata` on `Slackware current` (the parent) and let a child (`FreeBSD 13.1-RELEASE-p2 Libvirt VM`) connected to it. While the two `netdata` was running, I was working with other software (Firefox/Konsole/Visual Studio Code + go.d.plugin) on the same host.

ktsaou commented:
I this this is related to sqlite. Let's wait for @stelfrag to see it.

@thiagoftsm, I am preparing some changes to cleanup streaming code a bit. I found a lot of issues and I decided to fix them tonight. My changes are on top of this PR.

Since we are investigating this issue and we can't merge, I would prefer to commit them to this PR.
What do you think? Can you help to test them with this PR?

thiagoftsm commented:
> I this this is related to sqlite. Let's wait for @stelfrag to see it.
> 
> @thiagoftsm, I am preparing some changes to cleanup streaming code a bit. I found a lot of issues and I decided to fix them tonight. My changes are on top of this PR.
> 
> Since we are investigating this issue and we can't merge, I would prefer to commit them to this PR. What do you think? Can you help to test them with this PR?

I agree 100% with you, I am waiting for your commits to retest. :)

ktsaou commented:
Pushed it.

This should improve streaming performance on children quite a lot, mainly because:

1. I cleaned up the code that was doing a tone of atomic operations - now only 2 are made (one for rrdhost flags and one for rrdset flags) for each `rrdset_done_push()`.
2. The mutex that was doing all operations to push metrics was exclusively locked for the entire host, so only one thread at a time could use it. Now each thread has its own buffer and then the mutex is applied only for protecting the shared compression structures and the circular buffer.

I also moved all sender related members of RRDHOST to the sender state structure.

thiagoftsm commented:
> Pushed it.
> 
> This should improve streaming performance on children quite a lot, mainly because:

Thank you to explain everything. On my end:

1. I will compile this PR on different hosts and let it running.
2. While I am testing I will review the code.
3. Finally I will run `coverity` one more time on this PR.

I will bring updates as soon I finish the three steps. :handshake:

thiagoftsm commented on streaming/rrdpush.c:
```diff
@@ -130,40 +130,35 @@ unsigned int remote_clock_resync_iterations = 60;
-                rrdset_flag_clear(st, RRDSET_FLAG_UPSTREAM_IGNORE);
+
+        else if (flags & RRDSET_FLAG_ANOMALY_DETECTION) {
+            if(ml_streaming_enabled())
```
Hello @ktsaou ,

We have this new coverity report:

```sh
*** CID 358260:  Control flow issues  (DEADCODE)
/streaming/rrdpush.c: 148 in should_send_chart_matching()
142     
143             // Do not stream anomaly rates charts.
144             if (unlikely(flags & RRDSET_FLAG_ANOMALY_RATE_CHART))
145                 rrdset_flag_set(st, RRDSET_FLAG_UPSTREAM_IGNORE);
146     
147             else if (flags & RRDSET_FLAG_ANOMALY_DETECTION) {
>>>     CID 358260:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach this statement: "if (ml_streaming_enabled())...".
148                 if(ml_streaming_enabled())
149                     rrdset_flag_set(st, RRDSET_FLAG_UPSTREAM_SEND);
150                 else
151                     rrdset_flag_set(st, RRDSET_FLAG_UPSTREAM_IGNORE);
152             }
153             else if(simple_pattern_matches(host->rrdpush_send_charts_matching, rrdset_id(st)) ||
```

thiagoftsm commented:
@ktsaou and @stelfrag ,

After stream to be rewritten, I could finally run parent more than 3 hours without crashes  (Thanks, and of course congratulations @ktsaou !). :)

During the tests I used the following environment:

| Host | Distribution | Is the parent ? | 
|--------|----------------|----------------------|
| x86_64 | Slackware current | yes |
|x86_64 libvirt VM | FreeBSD 13.1-RELEASE-p2 |  no |
|armvl7 | Raspberry pi 4 OS | no |

and everything worked as expected.

On the other hand, I can confirm that PR cannot be compiled on CentOS 7.9 with the same error that CI is reporting:

```sh
[/home/thiago/Netdata/netdata]# make -j1 
streaming/sender.c: In function 'sender_commit':
streaming/sender.c:146:10: error: 'struct sender_state' has no member named 'overflow'
         s->overflow = 1;
          ^
make[2]: *** [streaming/sender.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
 FAILED  ''

 ABORTED  Failed to build Netdata. 
```

, please, take a look on this.

ktsaou commented:
fixed the leftover `s->overflow = 1;`

thiagoftsm commented:
> fixed the leftover `s->overflow = 1;`

Thank you, I could compile on `CentOS 7.9` and I restart tests after this. :)

thiagoftsm commented:
@stelfrag and @ktsaou ,

I had the issue again today:

```sh
==31426==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7efe0a6298c0 at pc 0x00000046f56b bp 0x7efe0a628fd0 sp 0x7efe0a628790
WRITE of size 56 at 0x7efe0a6298c0 thread T48
    #0 0x46f56a in unpoison_tm(void*, __sanitizer::__sanitizer_tm*) (/usr/sbin/netdata+0x46f56a)
    #1 0x478a40 in __interceptor_localtime_r (/usr/sbin/netdata+0x478a40)
    #2 0x567aa1 in log_date libnetdata/log/log.c:464
    #3 0x567aa1 in info_int libnetdata/log/log.c:769
    #4 0x9d2c9b in release_statement database/sqlite/sqlite_functions.c:138
    #5 0x7efe2978ae20 in __GI___nptl_deallocate_tsd (/lib64/libc.so.6+0x88e20)
    #6 0x7efe2978de44 in start_thread (/lib64/libc.so.6+0x8be44)
    #7 0x7efe298189eb in clone3 (/lib64/libc.so.6+0x1169eb)

Address 0x7efe0a6298c0 is located in stack of thread T48 at offset 64 in frame
    #0 0x567938 in info_int libnetdata/log/log.c:751

  This frame has 4 object(s):
    [32, 40) 't' (line 460)
    [64, 120) 'tmbuf' (line 461) <== Memory access at offset 64 is inside this variable
    [160, 184) 'args' (line 752)
    [224, 250) 'date' (line 768)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
Thread T48 created by T46 here:
    #0 0x43f5e6 in pthread_create (/usr/sbin/netdata+0x43f5e6)
    #1 0x7efe29add618 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/usr/lib64/libstdc++.so.6+0xdd618)
    #2 0x91b6fa in ml_new_host ml/ml.cc:55
    #3 0x9b35c8 in rrdhost_create database/rrdhost.c:544
    #4 0x9b4934 in rrdhost_find_or_create database/rrdhost.c:687
    #5 0xbec99a in rrdpush_receive streaming/receiver.c:548
    #6 0xbef020 in rrdpush_receiver_thread streaming/receiver.c:825
    #7 0x586bf8 in thread_start libnetdata/threads/threads.c:201
    #8 0x7efe2978dfe9 in start_thread (/lib64/libc.so.6+0x8bfe9)

Thread T46 created by T37 here:
    #0 0x43f5e6 in pthread_create (/usr/sbin/netdata+0x43f5e6)
    #1 0x587258 in netdata_thread_create libnetdata/threads/threads.c:215
    #2 0xbdcd92 in rrdpush_receiver_thread_spawn streaming/rrdpush.c:900
    #3 0xc08bb2 in web_client_process_request web/server/web_client.c:1469
    #4 0xc0fca0 in web_server_rcv_callback web/server/static/static-threaded.c:304
    #5 0x57ed7d in poll_process_tcp_read libnetdata/socket/socket.c:1493
    #6 0x57ed7d in poll_events libnetdata/socket/socket.c:1812
    #7 0xc0f609 in socket_listen_main_static_threaded_worker web/server/static/static-threaded.c:427
    #8 0x586bf8 in thread_start libnetdata/threads/threads.c:201
    #9 0x7efe2978dfe9 in start_thread (/lib64/libc.so.6+0x8bfe9)

Thread T37 created by T18 here:
    #0 0x43f5e6 in pthread_create (/usr/sbin/netdata+0x43f5e6)
    #1 0x587258 in netdata_thread_create libnetdata/threads/threads.c:215
    #2 0xc12413 in socket_listen_main_static_threaded web/server/static/static-threaded.c:540
    #3 0x586bf8 in thread_start libnetdata/threads/threads.c:201
    #4 0x7efe2978dfe9 in start_thread (/lib64/libc.so.6+0x8bfe9)

Thread T18 created by T0 here:
    #0 0x43f5e6 in pthread_create (/usr/sbin/netdata+0x43f5e6)
    #1 0x587258 in netdata_thread_create libnetdata/threads/threads.c:215
    #2 0x516f2f in main daemon/main.c:1507
    #3 0x7efe29725176 in __libc_start_call_main (/lib64/libc.so.6+0x23176)

```

thiagoftsm commented:
I can confirm last commit fix the issue with `streaming`.

underhood commented:
While this is definitely an awesome functionality that will add lot of usability to the product, I worry about security implications here a bit. Will this not expose netdata/system its on to the bugs in the plugins?

underhood commented:
Idea: I think there was a plan to do protobuf based streaming protocol? Should we think about building this on top of protobuf RPC capabilities?

underhood commented on collectors/plugins.d/README.md:
```diff
@@ -362,6 +364,80 @@ The `source` is an integer field that can have the following values:
+
+> FUNCTION transaction_id timeout "name and parameters of the function"
+
+The plugin is expected to parse and validate `name and parameters of the function`. Netdata allows users to edit this string, append more parameters or even change the ones the plugin originally exposed. To minimize the security risk, Netdata guarantees that only the characters shown above are accepted in function definitions, but still the plugin should carefully inspect the `name and parameters ...
```
`name and parameters of the function` - how it should look? Documentation seems unclear to me. If it is supposed to be arbitrary string to be interpreted at plugins own free will shouldn't we call it something like `command`?
If the format is enforced to `"fnc_name [parameters]*"` shouldnt the function name be a separate parameter on the line? (not together in string with params)

ktsaou replied:
This should be a contract between UI and plugins. This is why I have not explained it more currently.
I believe it should be something like:

```
FUNCTION [parameter] [parameter]
```

When each `parameter` should be `key[:value[,value]]`.

So, any `key` is accepted. But also `key:value` or even `key:value1,value2,value3`.

The way this works today, is that the plugin should register a function with the minimum number of parameters required to execute the function. For example, a plugin may say:

```
PostgresqlSlowQuery db:sales
```

But then the plugin could allow the user to provide additional parameters, like `table:revenue` or even `last:100`.

Based on the discussions we will have with the UI, we may also support a number parameters to help the UI be more performant, like `columns:col1,col2,col3,col4` to request fewer of the available columns, or even `sort:column:ascending rows:100` to have the plugin return the top 100 rows after sorting the table on `column` ascending.

So, all these remain to be fine decided.

ktsaou commented:
> Idea: I think there was a plan to do protobuf based streaming protocol? Should we think about building this on top of protobuf RPC capabilities?

We have to find the most efficient protocol for this. We are not ready yet to take this step. The key decision is how fast can a protocol be for metrics streaming. Today netdata peaks at about:

- 600k dimensions per second per node for the children, if they all come from a single plugin (e.g. go.d.plugin)
- 1.4M dimensions per second per node at a parent (not a proxy)
- and I estimate about 600k dimensions per second per node at a proxy (mainly because it the single threaded performance, like the child)

If protobuf cannot help us have better performance, I am not sure we should use it. So, it needs to be tested before doing it.

So, we need to test this.

ktsaou commented:
Merged it!

Thank you all guys! A new era begins...
